src/modules/kafka/kfk.c

```diff
@@ -0,0 +1,1175 @@
+
+/**
+ * \brief list of configuration properties.
+ */
+typedef struct kfk_conf_s {
+	param_t *attrs; /**< parsed attributes from configuration parameter. */
+	char *spec; /**< original string of configuration. */
+	kfk_conf_node_t *property; /**< list of configuration properties. */
+} kfk_conf_t;
+
+/**
+ * \brief data type for a topic.
+ *
+ * This is an element in a topic list.
+ */
+typedef struct kfk_topic_s {
+	str *topic_name; /**< Name of the topic. */
+	rd_kafka_topic_t *rd_topic; /**< rd kafkfa topic structure. */
+	param_t *attrs; /**< parsed attributes for topic configuration. */
+	char *spec; /**< original string for topic configuration. */
+	kfk_conf_node_t *property; /**< list of configuration properties for a topic. */
+	struct kfk_topic_s *next; /**< Next element in topic list. */
+} kfk_topic_t;
+
+/**
+ * \brief stats about a topic.
+ */
+typedef struct kfk_stats_s {
+	str *topic_name; /**< Name of the topic, or NULL for general statistics. */
+	uint64_t total; /**< Total number of messages sent. */
+	uint64_t error; /**< Number of failed messages to sent. */
+	struct kfk_stats_s *next; /**< Next element in stats list. */
+} kfk_stats_t;
+
+/* Static variables. */
+static rd_kafka_conf_t *rk_conf = NULL;  /* Configuration object */
+static rd_kafka_t *rk = NULL; /* Producer instance handle */
+static kfk_conf_t *kfk_conf = NULL; /* List for Kafka configuration properties. */
+static kfk_topic_t *kfk_topic = NULL; /* List for Kafka topics. */
+
+#define ERRSTR_LEN 512 /**< Length of internal buffer for errors. */
+static char errstr[ERRSTR_LEN]; /* librdkafka API error reporting buffer */
+gen_lock_t *stats_lock = NULL; /**< Lock to protect shared statistics data. */
+
+/**
+ * \brief Total statistics
+ *
+ * First node (mandatory) is the general one with NULL topic.
+ * Next nodes are topic dependant ones and are optional.
+ * This way because general node is created in kfk_stats_init in mod_init is
+ * shared among every Kamailio process.
+ */
+static kfk_stats_t *stats_general;
+
+/* Static functions. */
+static void kfk_conf_free(kfk_conf_t *kconf);
+static void kfk_topic_free(kfk_topic_t *ktopic);
+static int kfk_conf_configure();
+static int kfk_topic_list_configure();
+static int kfk_topic_exist(str *topic_name);
+static rd_kafka_topic_t* kfk_topic_get(str *topic_name);
+static int kfk_stats_add(const char *topic, rd_kafka_resp_err_t err);
+static void kfk_stats_topic_free(kfk_stats_t *st_topic);
+
+/**
+ * \name Logging_macros from /usr/include/sys/syslog.h
+ */
+/*@{*/
+#define	LOG_EMERG	0	/* system is unusable */
+#define	LOG_ALERT	1	/* action must be taken immediately */
+#define	LOG_CRIT	2	/* critical conditions */
+#define	LOG_ERR		3	/* error conditions */
+#define	LOG_WARNING	4	/* warning conditions */
+#define	LOG_NOTICE	5	/* normal but significant condition */
+#define	LOG_INFO	6	/* informational */
+#define	LOG_DEBUG	7	/* debug-level messages */
+/*@}*/
+
+/**
+ * \brief Kafka logger callback (optional)
```

Comment:
You mention that this callback is optional, but you set it always - maybe adapt the comment.